python/tests/test_pipeline.py

```diff
@@ -0,0 +1,436 @@
+# limitations under the License.
+#
+
+from typing import Any, Dict, List, Optional, Tuple, Type, Union
+
+import numpy as np
+import pyspark.ml.functions as MLF
+import pyspark.sql.functions as SQLF
+import pytest
+from pyspark.ml import Pipeline as SparkPipeline
+from pyspark.ml.classification import LogisticRegression as SparkLogisticRegression
+from pyspark.ml.classification import (
+    LogisticRegressionModel as SparkLogisticRegressionModel,
+)
+from pyspark.ml.feature import HashingTF, Tokenizer, VectorAssembler
+from pyspark.ml.tuning import CrossValidator as SparkCrossValidator
+from pyspark.sql import DataFrame, SparkSession
+
+from spark_rapids_ml.classification import LogisticRegression as GPULogisticRegression
+from spark_rapids_ml.classification import (
+    LogisticRegressionModel as GPULogisticRegressionModel,
+)
+from spark_rapids_ml.core import _CumlEstimator
+from spark_rapids_ml.pipeline import NoOpTransformer, Pipeline
+from spark_rapids_ml.tuning import CrossValidator
+
+from .utils import array_equal, create_pyspark_dataframe, make_classification_dataset
+
+
+def create_toy_dataframe(
+    spark: SparkSession, all_scalar_columns: bool = True
+) -> Tuple[DataFrame, List[str], str]:
+    import numpy as np
+
+    X = np.array(
+        [
+            [1.0, 2.0],
+            [1.0, 3.0],
+            [2.0, 1.0],
+            [3.0, 1.0],
+        ]
+    )
+
+    y = np.array(
+        [
+            1.0,
+            1.0,
+            0.0,
+            0.0,
+        ]
+    )
+
+    (df, feature_cols, label_col) = create_pyspark_dataframe(
+        spark, "multi_cols", np.dtype("float32"), X, y
+    )
+
+    if all_scalar_columns is False:
+        col_name = feature_cols[0]
+
+        df = df.withColumn("c0_vec", MLF.array_to_vector(SQLF.array([col_name])))
+        df = df.drop(col_name)
+        df = df.withColumnRenamed("c0_vec", col_name)
+
+    return (df, feature_cols, label_col)  # type: ignore
+
+
+@pytest.mark.parametrize(
+    "PipelineEst,Est,Model",
+    [
+        (Pipeline, GPULogisticRegression, GPULogisticRegressionModel),
+        (Pipeline, SparkLogisticRegression, SparkLogisticRegressionModel),
+        (SparkPipeline, GPULogisticRegression, GPULogisticRegressionModel),
+        (SparkPipeline, SparkLogisticRegression, SparkLogisticRegressionModel),
+    ],
+)
+def test_example(
+    PipelineEst: Type,
+    Est: Type,
+    Model: Type,
+    caplog: pytest.LogCaptureFixture,
+    all_scalar_colum: bool = True,
+) -> None:
+
+    from .conftest import _spark
+
+    (df, input_cols, label_col) = create_toy_dataframe(_spark, all_scalar_colum)
+    assembler = VectorAssembler(inputCols=input_cols, outputCol="features")
+
+    algo_params = {
+        "featuresCol": "features",
+        "labelCol": label_col,
+        "maxIter": 10,
+        "regParam": 0.001,
+    }
+    est = Est(**algo_params)
+
+    pipeline = PipelineEst(stages=[assembler, est])
```

Comment:
Any tests with more than two stages?   Pipelines can be very complicated DAGs in the general case through column references to preceding stage outputs.   See comments above.

Reply:
Make sense. Now PR revised to handle two stages only. 
Need to investigage the case if more than two stages.